mixer/pkg/protobuf/yaml/dynamic/encoder.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamic
+
+import (
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+
+	"istio.io/istio/mixer/pkg/attribute"
+)
+
+type (
+	// Encoder transforms yaml that represents protobuf data into []byte
+	// The yaml representation may have dynamic content
+	Encoder interface {
+		Encode(bag attribute.Bag, ba []byte) ([]byte, error)
+	}
+
+	messageEncoder struct {
+		// skipEncodeLength skip encoding length of the message in the output
+		// should be true only for top level message.
+		skipEncodeLength bool
+
+		// fields of the message.
+		fields []*field
+	}
+
+	field struct {
+		// proto key  -- EncodeVarInt ((field_number << 3) | wire_type)
+		protoKey []byte
+		// if packed this is set to true
+		packed bool
+		// encoder is needed if encodedData is not available.
+		// packed fields have a list of encoders.
+		encoder []Encoder
+
+		// number fields are sorted by field number.
+		number int
+		// name for debug.
+		name string
+	}
+)
+
+func extendSlice(ba []byte, n int) []byte {
+	for k := 0; k < n; k++ {
+		ba = append(ba, 0xff)
```

Comment:
consider comment on what 0xff is

Reply:
made it `0x0`. I had it set to `0xff` for debugging earlier.